Examples/Io/Performance/ActsExamples/Io/Performance/TrackFinderNTupleWriter.cpp

```diff
@@ -0,0 +1,295 @@
+// This file is part of the ACTS project.
+//
+// Copyright (C) 2016 CERN for the benefit of the ACTS project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "ActsExamples/Io/Performance/TrackFinderNTupleWriter.hpp"
+
+#include "Acts/Definitions/Units.hpp"
+#include "ActsExamples/EventData/Index.hpp"
+#include "ActsExamples/EventData/SimHit.hpp"
+#include "ActsExamples/EventData/SimParticle.hpp"
+#include "ActsExamples/EventData/TruthMatching.hpp"
+#include "ActsExamples/Framework/AlgorithmContext.hpp"
+#include "ActsExamples/Framework/DataHandle.hpp"
+#include "ActsExamples/Utilities/Range.hpp"
+#include "ActsExamples/Validation/TrackClassification.hpp"
+#include "ActsFatras/EventData/Barcode.hpp"
+#include "ActsFatras/EventData/Particle.hpp"
+
+#include <cstddef>
+#include <cstdint>
+#include <mutex>
+#include <stdexcept>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include <RtypesCore.h>
+#include <TFile.h>
+#include <TTree.h>
+
+struct ActsExamples::TrackFinderNTupleWriter::Impl {
+  Config cfg;
+
+  ReadDataHandle<SimParticleContainer> inputParticles;
+  ReadDataHandle<HitParticlesMap> inputMeasurementParticlesMap;
+  ReadDataHandle<TrackParticleMatching> inputTrackParticleMatching;
+
+  TFile* file = nullptr;
+
+  // per-track tree
+  TTree* trkTree = nullptr;
+  std::mutex trkMutex;
+  // track identification
+  ULong64_t trkEventId = 0;
+  ULong64_t trkTrackId = 0;
+  // track content
+  // number of hits on track
+  UShort_t trkNumHits = 0;
+  // number of particles contained in the track
+  UShort_t trkNumParticles = 0;
+  // track particle content; for each contributing particle, largest first
+  std::vector<ULong64_t> trkParticleId;
+  // total number of hits generated by this particle
+  std::vector<UShort_t> trkParticleNumHitsTotal;
+  // number of hits within this track
+  std::vector<UShort_t> trkParticleNumHitsOnTrack;
+
+  // per-particle tree
+  TTree* prtTree = nullptr;
+  std::mutex prtMutex;
+  // particle identification
+  ULong64_t prtEventId = 0;
+  ULong64_t prtParticleId = 0;
+  Int_t prtParticleType = 0;
+  // particle kinematics
+  // vertex position in mm
+  float prtVx = 0, prtVy = 0, prtVz = 0;
+  // vertex time in ns
+  float prtVt = 0;
+  // particle momentum at production in GeV
+  float prtPx = 0, prtPy = 0, prtPz = 0;
+  // particle mass in GeV
+  float prtM = 0;
+  // particle charge in e
+  float prtQ = 0;
+  // particle reconstruction
+  UShort_t prtNumHits = 0;  // number of hits for this particle
+  UShort_t prtNumTracks =
+      0;  // number of tracks this particle was reconstructed in
+  UShort_t prtNumTracksMajority =
+      0;  // number of tracks reconstructed as majority
+  // extra logger reference for the logging macros
+  const Acts::Logger& _logger;
```

Comment:
If you just call this variable `logger`, I think you can even skip the separate `logger()` function.